File: java/src/main/java/nl/craftsmen/workshops/reactivemeetup/domain/railway/StationaryTrainSimulation.java
package nl.craftsmen.workshops.reactivemeetup.domain.railway;

import java.util.concurrent.TimeUnit;

import rx.Observable;

public class StationaryTrainSimulation implements TrainSimulation {
	
	private final LatLong position;
	
	private final double stationaryTime;
	
	public StationaryTrainSimulation(LatLong position, double stationaryTime) {
		this.position = position;
		this.stationaryTime = stationaryTime;
	}

	@Override
	public Observable<TrainMetrics> trainMetrics$(TrainSimulationParameters parameters, long startTime) {
		double tickDelay = 1000.0 / parameters.getTickFrequency();
		
		int requiredNumberOfFrames = (int) Math.ceil(stationaryTime * 1000 / (tickDelay * parameters.getTimeDilation()));
		
		return Observable.interval((long) Math.round(tickDelay), TimeUnit.MILLISECONDS)
			.take(requiredNumberOfFrames)
			.map((frameIndex) -> {
				double elapsedTime = ((frameIndex + 1) * tickDelay) / 1000.0 * parameters.getTimeDilation();
				
				return new TrainMetrics(parameters.getTrainId(), startTime + (long)(elapsedTime * 1000), position);
			});
	}

}

File: javascript/src/domain/railway/stationary-train-simulation.js
const Rx = require('rxjs/Rx');
const TrainMetrics = require('./train-metrics');

module.exports = class StationaryTrainSimulation {

	constructor(position, stationaryTime) {
		this.position = position;
		this.stationaryTime = stationaryTime;
	}

	trainMetrics$(parameters, startTime) {
				
		const tickDelay = 1000.0 / parameters.getTickFrequency();
		
		const requiredNumberOfFrames = Math.ceil(this.stationaryTime * 1000 / (tickDelay * parameters.getTimeDilation()));
		
		return Rx.Observable.interval(Math.round(tickDelay))
			.take(requiredNumberOfFrames)
			.map((frameIndex) => {
				const elapsedTime = ((frameIndex + 1) * tickDelay) / 1000.0 * parameters.getTimeDilation();
				
				return new TrainMetrics(parameters.getTrainId(), startTime + Math.floor(elapsedTime * 1000), this.position);
			});
	}
}	
